standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java

```diff
@@ -263,26 +235,124 @@ public String toString() {
     public Object value;
     public boolean isReverseOrder = false;
-    private static final String PARAM_PREFIX = "hive_filter_param_";
 
     @Override
     protected void accept(TreeVisitor visitor) throws MetaException {
       visitor.visit(this);
     }
 
     @Override
-    public void generateJDOFilter(Configuration conf, Map<String, Object> params,
-                                  FilterBuilder filterBuilder, List<FieldSchema> partitionKeys) throws MetaException {
+    public String toString() {
+      return "LeafNode{" +
+          "keyName='" + keyName + '\'' +
+          ", operator='" + operator + '\'' +
+          ", value=" + value +
+          (isReverseOrder ? ", isReverseOrder=true" : "") +
+          '}';
+    }
+
+    /**
+     * Get partition column index in the table partition column list that
+     * corresponds to the key that is being filtered on by this tree node.
+     * @param partitionKeys list of partition keys.
+     * @param filterBuilder filter builder used to report error, if any.
+     * @return The index.
+     */
+    public static int getPartColIndexForFilter(String partitionKeyName,
+        List<FieldSchema> partitionKeys, FilterBuilder filterBuilder) throws MetaException {
+      assert (partitionKeys.size() > 0);
+      int partitionColumnIndex;
+      for (partitionColumnIndex = 0; partitionColumnIndex < partitionKeys.size();
+           ++partitionColumnIndex) {
+        if (partitionKeys.get(partitionColumnIndex).getName().equalsIgnoreCase(partitionKeyName)) {
+          break;
+        }
+      }
+      if( partitionColumnIndex == partitionKeys.size()) {
+        filterBuilder.setError("Specified key <" + partitionKeyName +
+            "> is not a partitioning key for the table");
+        return -1;
+      }
+
+      return partitionColumnIndex;
+    }
+  }
+
+  /**
+   * Generate the JDOQL filter for the given expression tree
+   */
+  public static class JDOFilterGenerator extends TreeVisitor {
+
+    private static final String PARAM_PREFIX = "hive_filter_param_";
+
+    private Configuration conf;
+    private List<FieldSchema> partitionKeys;
+    // the filter builder to append to.
+    private FilterBuilder filterBuilder;
+    // the input map which is updated with the the parameterized values.
+    // Keys are the parameter names and values are the parameter values
+    private Map<String, Object> params;
+    private boolean onParsing = false;
+    private String keyName;
+    private Object value;
+    private Operator operator;
+    private boolean isReverseOrder;
+
+    public JDOFilterGenerator(Configuration conf, List<FieldSchema> partitionKeys,
+        FilterBuilder filterBuilder, Map<String, Object> params) {
+      this.conf = conf;
+      this.partitionKeys = partitionKeys;
+      this.filterBuilder = filterBuilder;
+      this.params = params;
+    }
+
+    private void beforeParsing() throws MetaException {
+      if (!onParsing && !filterBuilder.getFilter().isEmpty()) {
```

Comment:
do we need `onParsing` , isn't it enough to check the `filter`?

Reply:
won't `!filterBuilder.getFilter().isEmpty()` cover that?

Reply:
The `filterBuilder` could not be empty when begin parsing, e.g, https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L4905-L4917